### PR TITLE
prettierd: init at 0.23.4

### PR DESCRIPTION
**Pull request by NotAShelf**
###### Description of changes

Continuation of #231712. Cleans up the derivation and gets rid of the IDFs, adds the nodejs wrapper and a list of arguments to be passed to the daemon using `wrapProgram`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -3696,6 +3696,8 @@ with pkgs;
 
   platformsh = callPackage ../misc/platformsh { };
 
+  prettierd = callPackage ../development/tools/prettierd { };
```
Comment (fufexan) [inline pkgs/top-level/all-packages.nix:3699]: Line 11594 might be more appropriate.

File: pkgs/development/tools/prettierd/default.nix [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  postInstall = ''
+    wrapProgram "$out/bin/prettierd" --prefix PATH : "${nodejs}/bin" \
+      ${lib.concatStringsSep "\n" (map (a: "--add-flags ${builtins.toJSON a}") extraFlags)}
+  '';
+
+  doDist = false;
+
+  meta = with lib; {
+    mainProgram = "prettierd";
+    description = "prettier, as a daemon, for improved formatting speed.";
```
Comment (fufexan) [inline pkgs/development/tools/prettierd/default.nix:32]: Capitalize Prettier, remove the period.

File: pkgs/development/tools/prettierd/default.nix [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  '';
+
+  postInstall = ''
+    wrapProgram "$out/bin/prettierd" --prefix PATH : "${nodejs}/bin" \
+      ${lib.concatStringsSep "\n" (map (a: "--add-flags ${builtins.toJSON a}") extraFlags)}
+  '';
+
+  doDist = false;
+
+  meta = with lib; {
```
Comment (fufexan) [inline pkgs/development/tools/prettierd/default.nix:30]: `platform`?

File: pkgs/development/tools/prettierd/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
+{ lib , mkYarnPackage , fetchFromGitHub , nodejs , makeWrapper , extraFlags ? [ ] , ...
```
Comment (fufexan) [inline pkgs/development/tools/prettierd/default.nix:1]: Each entry should go on a new line.

**Review by fufexan — CHANGES_REQUESTED** [review]

**Review by fufexan — APPROVED** [review]

File: maintainers/maintainer-list.nix
Comment (emilylange) [inline maintainers/maintainer-list.nix:1]: Please split this change into a separate commit titled `maintainers: update NotAShelf`

Meaning two commits in total:
1. `maintainers: update NotAShelf`
1. `prettierd: init at 0.23.4`

File: pkgs/development/tools/prettierd/default.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
+{ lib
+, mkYarnPackage
+, fetchFromGitHub
+, nodejs
+, makeWrapper
+, extraFlags ? [ ]
+, ...
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:7]: Not needed

```suggestion
```

File: pkgs/development/tools/prettierd/default.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  '';
+
+  doDist = false;
+
+  meta = with lib; {
+    description = "Prettier, as a daemon, for improved formatting speed";
+    homepage = "https://github.com/fsouza/prettierd";
+    license = licenses.isc;
+    changelog = "https://github.com/fsouza/prettierd/blob/main/CHANGELOG.md";
+    mainProgram = "prettierd";
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:41]:
```suggestion
```

Defaults to `pname`

File: pkgs/development/tools/prettierd/default.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      ${lib.concatStringsSep "\n" (map (a: "--add-flags ${builtins.toJSON a}") extraFlags)}
+  '';
+
+  doDist = false;
+
+  meta = with lib; {
+    description = "Prettier, as a daemon, for improved formatting speed";
+    homepage = "https://github.com/fsouza/prettierd";
+    license = licenses.isc;
+    changelog = "https://github.com/fsouza/prettierd/blob/main/CHANGELOG.md";
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:40]:
```suggestion
    changelog = "https://github.com/fsouza/prettierd/blob/${srv.rev}/CHANGELOG.md";
```

File: pkgs/development/tools/prettierd/default.nix [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    mainProgram = "prettierd";
+    platforms = lib.platforms.linux;
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:42]: You have a `with lib;` in that scope, thus you can omit `lib` here:

```suggestion
    platforms = platforms.linux;
```

File: pkgs/development/tools/prettierd/default.nix [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    maintainers = with maintainers; [ NotAShelf n3oney ];
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:43]: `n3oney` is not in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix and even that would be the case, please ask them first :)

```suggestion
    maintainers = with maintainers; [ NotAShelf ];
```

File: pkgs/development/tools/prettierd/default.nix [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  buildPhase = ''
+    runHook preBuild
+    yarn prepare
+    yarn build
+    runHook postBuild
+  '';
+
+  postInstall = ''
+    wrapProgram "$out/bin/prettierd" --prefix PATH : "${nodejs}/bin" \
+      ${lib.concatStringsSep "\n" (map (a: "--add-flags ${builtins.toJSON a}") extraFlags)}
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:31]: Can you explain why `--add-flags` is needed here?
I couldn't spot a mention of it in https://github.com/fsouza/prettierd

Thanks!

**Review by emilylange — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (NotAShelf) [reply maintainers/maintainer-list.nix:1]: My understanding was that multiple commits in one PR were frowned upon, will change.

File: pkgs/development/tools/prettierd/default.nix (hunk 9 above)
Comment (NotAShelf) [reply pkgs/development/tools/prettierd/default.nix:43]: This PR has been done in collaboration with n3oney, we are in communication on Discord, and I added him as a maintainer within his knowledge.

I forgot that the PR that added him to the maintainers list was never merged, will update the maintainers list unless @n3oney has an objection.

File: pkgs/development/tools/prettierd/default.nix (hunk 10 above)
Comment (NotAShelf) [reply pkgs/development/tools/prettierd/default.nix:31]: --add-flags is to wrap the program with flags if the user so chooses to i.e
```
prettierd {
    extraFlags = [ "--flag1" "--boolflag=true" ]; etc. 
}
```

Since prettierd is meant to run as a daemon, I assumed the user might want to pass those flags to a potential systemd service or a container without hassle.

File: pkgs/development/tools/prettierd/default.nix (hunk 9 above)
Comment (n3oney) [reply pkgs/development/tools/prettierd/default.nix:43]: I’m fine with it

File: maintainers/maintainer-list.nix
Comment (emilylange) [reply maintainers/maintainer-list.nix:1]: No, every distinct change should be in its own separate commit.
Think atomic commits.

But fixups and such, should be squashed.

Excerpt from https://nixos.org/manual/nixpkgs/stable/#submitting-changes-making-patches, which is linked to in https://github.com/NixOS/nixpkgs/blob/87ef73d4e3c7c4a64dc0f8a653f7d7886cc6b459/CONTRIBUTING.md

>  Make commits of logical units. 

And from [the manual's section regarding `maintainer-list.nix`](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers):

> There is no restriction to becoming a maintainer, just add yourself to that list in a separate commit titled "maintainers: add alice" in the same pull request, and reference maintainers with `maintainers = with lib.maintainers; [ alice bob ]`.

File: pkgs/development/tools/prettierd/default.nix (hunk 10 above)
Comment (emilylange) [reply pkgs/development/tools/prettierd/default.nix:31]: Oh whoops, yeah, thanks.

I don't know why I tried to look in `prettierd`'s documentation for that.

Though I do feel like this isn't the right place for that.

Summoning @drupol, who reviewed the previous attempt at #231712

File: maintainers/maintainer-list.nix
Comment (NotAShelf) [reply maintainers/maintainer-list.nix:1]: I understand, thank you for the clarification!

**Comment by fufexan** [comment]
I think it would be good to have neoney co-sign the commits.

**Comment by NotAShelf** [comment]
> I think it would be good to have neoney co-sign the commits. (The one adding him to the maintainers list, and the prettierd init one)

Good idea, but I frankly have no idea how to have him co-sign the commits without screwing up the git history. Besides, I am responsible for the code contributed in this PR.

**Comment by drupol** [comment]
@n3oney Are you ok with this (being added as maintainer) ? Could you please verify if the informations are correct in the relevant commit?

**Comment by fufexan** [comment]
You can do `git rebase --interactive HEAD~3` and `reword` the last 2 commits, then add `Co-authored-by: name <mail@example.com>` on a new line. The name should be the one he uses on github (so Michał), same for the email.

**Comment by fufexan** [comment]
@drupol he responded [here](https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251943001). I still think making him a co-author is a good idea.

**Review by drupol — CHANGES_REQUESTED** [review]
Please sort the maintainer list (see the broken CI check there).

**Comment by emilylange** [comment]
And you marked most of my suggestions as "resolved" without changing the things requested

**Comment by drupol** [comment]
> And you marked most of my suggestions as "resolved" without changing the things requested

?

**Comment by emilylange** [comment]
Was directed to @NotAShelf (sorry if that wasn't clear).

Ref:

- https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251291219
- https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251291513
- https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251291648
- https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251292123

**Comment by NotAShelf** [comment]
Must've screwed up in the git history somehow. Will take another look.

**Comment by NotAShelf** [comment]
It was indeed broken, thanks for pointing it out!

With that fixed, should be all. @emilylange am I missing anything else?

**Comment by emilylange** [comment]
Thanks @NotAShelf! :)

Looks good to me (though I don't do a lot of node packaging).

@drupol: What do you think of the `wrapProgram` at https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251297217?

**Review by drupol — APPROVED** [review]
LGTM now :)

**Comment by NotAShelf** [comment]
Draft?

**Comment by NotAShelf** [comment]
@emilylange is there a specific reason for the conversion to draft?

**Comment by emilylange** [comment]
Yeah, sorry.
This PR isn't ready to merge (because it uses import-from-derivation [IFD], which aren't allowed in nixpkgs).
See the ofborg eval error.

I intended to send you a patch yesterday, but ran out of time.

I marked this as draft, so no one accidentally merges this.

Will try to finish that patch later.

**Review by emilylange — CHANGES_REQUESTED** [review]
Removed IFD and fixed eval:

```patch
diff --git a/pkgs/development/tools/prettierd/default.nix b/pkgs/development/tools/prettierd/default.nix
index e5615d9725e..0f149f04f19 100644
--- a/pkgs/development/tools/prettierd/default.nix
+++ b/pkgs/development/tools/prettierd/default.nix
@@ -1,10 +1,9 @@
 { lib
 , mkYarnPackage
+, fetchYarnDeps
 , fetchFromGitHub
-, nodejs
 , makeWrapper
 , extraFlags ? [ ]
-,
 }:
 mkYarnPackage rec {
   pname = "prettierd";
@@ -17,18 +16,22 @@ mkYarnPackage rec {
     hash = "sha256-GTukjkA/53N9ICdfCJr5HAqhdL5T0pth6zAk8Fu/cis=";
   };
 
+  offlineCache = fetchYarnDeps {
+    yarnLock = src + "/yarn.lock";
+    hash = "sha256-32wMwkVgO5DQuROWnujVGNeCAUq1D6jJurecsD2ROOU=";
+  };
+
+  packageJSON = ./package.json;
+
   nativeBuildInputs = [ makeWrapper ];
 
   buildPhase = ''
     runHook preBuild
-    yarn prepare
-    yarn build
-    runHook postBuild
-  '';
 
-  postInstall = ''
-    wrapProgram "$out/bin/prettierd" --prefix PATH : "${nodejs}/bin" \
-      ${lib.concatStringsSep "\n" (map (a: "--add-flags ${builtins.toJSON a}") extraFlags)}
+    export HOME=$(mktemp -d)
+    yarn --offline build
+
+    runHook postBuild
   '';
 
   doDist = false;
diff --git a/pkgs/development/tools/prettierd/package.json b/pkgs/development/tools/prettierd/package.json
new file mode 100644
index 00000000000..8769bac80c7
--- /dev/null
+++ b/pkgs/development/tools/prettierd/package.json
@@ -0,0 +1,46 @@
+{
+  "name": "@fsouza/prettierd",
+  "version": "0.23.4",
+  "description": "prettier, as a daemon",
+  "bin": {
+    "prettierd": "./bin/prettierd"
+  },
+  "scripts": {
+    "prepare": "yarn build",
+    "build": "tsc -b",
+    "run": "node bin/prettierd",
+    "start": "node bin/prettierd start",
+    "prettier:check": "prettier --check .",
+    "prettier:fix": "prettier --write ."
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/fsouza/prettierd.git"
+  },
+  "author": "Francisco Souza",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/fsouza/prettierd/issues"
+  },
+  "homepage": "https://github.com/fsouza/prettierd",
+  "devDependencies": {
+    "@types/node": "^20.2.5",
+    "@types/prettier": "^2.7.2",
+    "typescript": "^5.0.4"
+  },
+  "dependencies": {
+    "core_d": "^5.0.1",
+    "nanolru": "^1.0.0",
+    "prettier": "^2.8.8"
+  },
+  "files": [
+    "bin",
+    "dist",
+    "LICENSE",
+    "README.md"
+  ],
+  "optionalDependencies": {
+    "@babel/parser": "^7.22.3",
+    "@typescript-eslint/typescript-estree": "^5.59.7"
+  }
+}
```

I am still not a fan of the `wrapProgram` you use in `postInstall`, as I'd argue, additional command line arguments should be exposed via some sort of module and not in the package definition itself.
But if you prefer it this way anyway, feel free to add it again.

Anyhow, thank you for your patience! :)

**Comment by NotAShelf** [comment]
I understand, thank you!

In hindsight, I think my wrapProgram should not have been there in the first place: the user is already at the liberty of wrapping the package on their own terms, should it be necessary.

**Comment by NotAShelf** [comment]
I'll apply the changes when I'm on my desktop.

**Comment by NotAShelf** [comment]
@emilylange I've removed the unnecessary extraFlags part of the wrapper, but the program still needs to be wrapped with nodejs in order for the daemon to be functional. That is why I've kept
```nix
postInstall = ''
   wrapProgram "$out/bin/prettierd" --prefix PATH : "${nodejs}/bin"
'';
  ```
  
Additionally, I am concerned about the maintenance of the package.json - does it mean that when we would like too update it, we need to clone the package somewhere, run npm install to generate a package.json and then copy it manually to the nixpkgs directory?

**Comment by emilylange** [comment]
@NotAShelf whoops, yeah, I guess.
Keep that `wrapProgram` then :+1: 



> Additionally, I am concerned about the maintenance of the package.json - does it mean that when we would like too update it (package.json), we need to get the package.json from source manually and then copy it to the nixpkgs directory?

Yes, unfortunately.
The `package.json` needs to be updated with each version bump/release.

As far as I am aware, that's just a limitation of `mkYarnPackage` when `import-from-derivation` is disabled :(

But just to be sure:
cc @NixOS/node

File: pkgs/development/tools/prettierd/default.nix [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildPhase = ''
+    runHook preBuild
+    export HOME=$(mktemp -d)
+    yarn --offline build
+
+    runHook postBuild
+    runHook postBuild
+  '';
```
Comment (emilylange) [inline pkgs/development/tools/prettierd/default.nix:36]:
```suggestion

    runHook postBuild
  '';
```

Same line twice (error when applying the patch I guess).

Remove one instance of `runHook postBuild` please :)

**Review by emilylange — CHANGES_REQUESTED** [review]

**Review by n3oney — APPROVED** [review]

**Review by n3oney — APPROVED** [review]

**Comment by NotAShelf** [comment]
I know we've previously stated that mainProgram defaults to `pname`, but it appears that it's trying to use `@fsouza/prettierd'` after the changes.

Should we provide `prettierd` as the mainProgram so that `nix run .#prettierd` is functional again?

**Comment by lilyinstarlight** [comment]
> As far as I am aware, that's just a limitation of `mkYarnPackage` when `import-from-derivation` is disabled :(

Yeah, unfortunately yarn2nix stuff requires that. We will hopefully have better yarn tooling soon (`buildYarnPackage` is next or so on our list), but for now it's either not use yarn2nix or vendor package.json iirc. I'm not at a computer to double check if there's a workaround

**Comment by lilyinstarlight** [comment]
So we could actually get rid of the `package.json`/IFD/`pname`/etc problems with a modest patch to yarn2nix it looks like:

```diff
diff --git a/pkgs/development/tools/prettierd/default.nix b/pkgs/development/tools/prettierd/default.nix
index 400639a543d..43c9dbe6e6d 100644
--- a/pkgs/development/tools/prettierd/default.nix
+++ b/pkgs/development/tools/prettierd/default.nix
@@ -22,8 +22,6 @@ mkYarnPackage rec {
     hash = "sha256-32wMwkVgO5DQuROWnujVGNeCAUq1D6jJurecsD2ROOU=";
   };
 
-  packageJSON = ./package.json;
-
   nativeBuildInputs = [ makeWrapper ];
 
   buildPhase = ''
diff --git a/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix b/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix
index e0bd54fd11a..854da527cba 100644
--- a/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix
+++ b/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix
@@ -267,7 +267,7 @@ in rec {
   }@attrs:
     let
       package = lib.importJSON packageJSON;
-      pname = package.name;
+      pname = attrs.pname or package.name;
       safeName = reformatPackageName pname;
       version = attrs.version or package.version;
       baseName = unlessNull name "${safeName}-${version}";
@@ -390,10 +390,11 @@ in rec {
 
       meta = {
         inherit (nodejs.meta) platforms;
-      } // lib.optionalAttrs (package ? description) { inherit (package) description; }
+      } // (attrs.meta or (
+           lib.optionalAttrs (package ? description) { inherit (package) description; }
         // lib.optionalAttrs (package ? homepage) { inherit (package) homepage; }
         // lib.optionalAttrs (package ? license) { license = getLicenseFromSpdxId package.license; }
-        // (attrs.meta or {});
+      ));
     });
 
   yarn2nix = mkYarnPackage {
```

I think I'll PR that diff since this is definitely not the only package that would benefit from using yarn2nix without IFD or vendoring (until we get new tooling and yarn2nix is no longer necessary)

**Comment by emilylange** [comment]
@NotAShelf 

> I know we've previously stated that mainProgram defaults to `pname`, but it appears that it's trying to use `@fsouza/prettierd'` after the changes.
> 
> Should we provide `prettierd` as the mainProgram so that `nix run .#prettierd` is functional again?

Good catch! TIL.
I wasn't aware that `mkYarnPackage` shadows the `pname`.

Please [undo my suggestion](https://github.com/NixOS/nixpkgs/pull/241339#discussion_r1251291513) and set `meta.mainProgram = "prettierd"` again.
Thank you!

---

@lilyinstarlight 

That would be awesome!

**Comment by NotAShelf** [comment]
All should be good now.
